coordinator/tributary/src/mempool.rs

```diff
@@ -118,27 +175,31 @@ impl<D: Db, T: Transaction> Mempool<D, T> {
   pub(crate) fn block(
     &mut self,
     blockchain_next_nonces: &HashMap<<Ristretto as Ciphersuite>::G, u32>,
-  ) -> Vec<T> {
+  ) -> Vec<Transaction<T>> {
     let mut res = vec![];
     for hash in self.txs.keys().cloned().collect::<Vec<_>>() {
       let tx = &self.txs[&hash];
+
       // Verify this hasn't gone stale
-      match tx.kind() {
-        TransactionKind::Signed(Signed { signer, nonce, .. }) => {
-          if blockchain_next_nonces[signer] > *nonce {
-            self.remove(&hash);
-            continue;
+      if let Transaction::Application(tx) =  tx {
+        match tx.kind() {
+          TransactionKind::Signed(Signed { signer, nonce, .. }) => {
+            if blockchain_next_nonces[signer] > *nonce {
+              self.remove(&hash);
+              continue;
+            }
           }
+          _ => panic!("non-signed transaction entered mempool"),
```

Comment:
Same commentary on above about handling unsigned TXs here. We just need to check they weren't already included.

Reply:
Don't we already check whether a given unsigned is already in chain when adding to the pool? why do we need to re check here?

Reply:
This block is checking if the transaction didn't go stale (included on chain). It has to be run on all TXs, not just application TXs. Unsigned TXs are stale if they were already included on chain.

Technically, if they're included on chain, they should be removed from the mempool and accordingly should never be present in the mempool if they're stale. This is distinct from Signed where another TX with the same nonce may enter, making the TX in our mempool stale, despite not being on-chain.

The reason I still want you to handle unsigned TXs here is:
1) It's assuming a lack of a race condition which is exceptionally easy to add here. I'd rather be robust.
2) You iterate over all TXs. If any unsigned TX is in the pool, this will panic. While you limited it to Application TXs, so it wouldn't panic on unsigned Tendermint TXs, application TXs are allowed to be unsigned. That means an unsigned application TX will hit this panic.